twenty-first/src/shared_math/b_field_element.rs

```diff
@@ -552,286 +552,251 @@ mod b_prime_field_element_test {
+                element_0 * (element_1 * element_2)
+            );
+        }
+    }
+
+    proptest! {
+        #[test]
+        fn multiplication_distributes_over_addition(
+            element_0: BFieldElement,
+            element_1: BFieldElement,
+            element_2: BFieldElement,
+        ) {
+            prop_assert_eq!(
+                element_0 * (element_1 + element_2),
+                element_0 * element_1 + element_0 * element_2
+            );
+        }
+    }
 
-        let neinneinnein = BFieldElement::new(999);
-        assert_eq!(999, neinneinnein.value());
+    proptest! {
+        #[test]
+        fn multiplication_with_inverse_gives_identity(bfe in arbitrary_non_zero_bfield_element()) {
+            prop_assert!((bfe.inverse() * bfe).is_one());
+        }
     }
 
-    #[test]
-    fn simple_generator_test() {
-        assert_eq!(BFieldElement::new(7), BFieldElement::generator());
-        assert!(BFieldElement::new(7)
-            .mod_pow(((1u128 << 64) - (1u128 << 32)) as u64)
-            .is_one());
-        assert!(!BFieldElement::new(7)
-            .mod_pow(((1u128 << 64) - (1u128 << 32)) as u64 / 2)
-            .is_one());
+    proptest! {
+        #[test]
+        fn division_by_self_gives_identity(bfe in arbitrary_non_zero_bfield_element()) {
+            prop_assert!((bfe / bfe).is_one());
+        }
+    }
+
+    proptest! {
+        #[test]
+        fn values_larger_than_modulus_are_handled_correctly(large_value in BFieldElement::P..) {
+            let bfe = BFieldElement::new(large_value);
+            let expected_value = large_value - BFieldElement::P;
+            prop_assert_eq!(expected_value, bfe.value());
+        }
     }
 
     #[test]
-    fn simple_lift_test() {
-        let zero: BFieldElement = BFieldElement::new(0);
-        assert!(zero.lift().is_zero());
-        assert!(!zero.lift().is_one());
+    fn display_test() {
+        let seven: BFieldElement = BFieldElement::new(7);
+        assert_eq!("7", format!("{seven}"));
 
-        let one: BFieldElement = BFieldElement::new(1);
-        assert!(!one.lift().is_zero());
-        assert!(one.lift().is_one());
+        let minus_one: BFieldElement = BFieldElement::new(BFieldElement::P - 1);
+        assert_eq!("-1", format!("{minus_one}"));
 
-        let five: BFieldElement = BFieldElement::new(5);
-        let five_lifted: XFieldElement = five.lift();
-        assert_eq!(Some(five), five_lifted.unlift());
+        let minus_fifteen: BFieldElement = BFieldElement::new(BFieldElement::P - 15);
+        assert_eq!("-15", format!("{minus_fifteen}"));
     }
 
     #[test]
-    fn lift_property_test() {
-        let elements: Vec<BFieldElement> = random_elements(100);
-        for element in elements {
-            assert_eq!(Some(element), element.lift().unlift());
-        }
+    fn zero_is_zero() {
+        let zero = BFieldElement::zero();
+        assert!(zero.is_zero());
     }
 
     #[test]
-    fn next_and_previous_test() {
```

Comment:
This test checked that `increment` and `decrement` handled the corner-cases correctly wrapping around the prime. I think your property-based tests don't hit those corner-cases.

Reply:
You are right, they corners are hit too rarely. I've added the tests `increment_overflow` and `decrement_underflow` for the two corner cases.

Reply:
Great. I don't like the names, though. Maybe `increment_wrap`, `decrement_wrap` instead? Underflow refers to a floating point becoming zero, not to a wrap around from zero to the highest possible value.

Reply:
I've made it even more explicit with `incrementing_max_value_wraps_around` and `decrementing_min_value_wraps_around`.